ons-spark/spark-app/spark_application_and_ui.md

```diff
@@ -0,0 +1,429 @@
+# Spark Application and UI
+
+This article introduces the structure of a Spark application and how the structure relates to the various pages within the application monitoring interface, the Spark UI. 
+
+The Spark UI is used to monitor the status and resource consumption of your Spark cluster and is the main tool for troubleshooting slow Spark code. 
+
+Understanding the structure of a Spark application helps to understand how Spark works so you can begin to think differently as you code to get the most out of Spark.
+
+## Spark Application Overview
+A Spark application has an underlining structure and learning about this structure will help us
+1. understand the difference between a narrow transformation, wide transformation and an action, 
+2. navigate around the Spark UI
+
+Below is a diagram of the Spark application hierarchy.
+
+- **Application** - a set of jobs managed by a single driver, e.g. on a Cloudera platform that is the CDSW session. An application is created when we connect to Spark with a `SparkSession` in PySpark or `spark_connect()` in sparklyr.
+- **Job** - a set of stages executed as a result of an *action*, e.g. `.count()`/`sdf_nrow()`. A job might consist of a number of *transformations*, but a job always finishes with an *action*.
+- **Stage** - a set of tasks that can be executed in parallel. Similar to how *actions* define a job boundary, a stage boundary is introduced by a *wide* transformation, such as `.groupby()`/`group_by()` or `.join()`/`left_join()`. A stage might consist of many *narrow* transformations, but a stage always finishes with a *wide* transformation or action.
+- **Task** - an individual unit of work assigned to a single core on an executor.
+
+
+![Diagram of Spark application showing partitioned DataFrame going through narrow and wide operations](../images/spark_app_diagram.png)
+
+
+Note that Jobs, Stages and Tasks are numbered starting from zero.
+
+Now look more closely at the DataFrame. On the left there is a Spark DataFrame which is split up into four sections, called *partitions*. Spark puts our data into partitions automatically so we don't have to decide how or when to do it. More details on exactly how Spark does this are in the **Partitions Article**. 
+
+The first operation to be applied to this DataFrame is a *narrow* transformation, e.g. summing two columns to create a new column. The second operation is again a narrow transformation. These two operations can be run in parallel, because each child partition has only one parent, indicated by the golden arrows.
```

Comment:
I think I'd put _adding_ rather than _summing_ here. Although they are exactly the same thing, people may get confused with summing values in a column, which is a wide transformation.